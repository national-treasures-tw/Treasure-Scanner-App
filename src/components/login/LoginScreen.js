// @flow
import React from 'react'
import {
  AlertIOS,
  View,
  Image,
  Text,
  StatusBar,
  TouchableHighlight,
  Keyboard
} from 'react-native'
import LoginForm from './LoginForm';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';
// import LoginButtons from './LoginButtons'
import { connect } from 'react-redux'
import styles from './styles/LoginScreenStyle'
import { Images, Metrics } from '../Themes';
import * as actions from '../../actions/actions';
import { bindActionCreators } from 'redux';

export class LoginScreen extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      values: {},
      profilePic: '',
      // name: '',
      // email: '',
      // password: '',
      // passwordConfirmation: '',
      // buildingId: 1,
      // apartmentId: 1,
      emailError: null,
      passwordError: null,
      visibleHeight: Metrics.screenHeight,
      topLogo: { width: Metrics.screenWidth },
      keyboardShowing: false
    }
    this.isAttempting = false
  }

  // componentDidMount() {
  //   StatusBar.setHidden(true, true);
  // }

  onLoginButtonPressed = () => {
    // const { email, password } = this.state
    // const { attemptLogin } = this.props
    // this.isAttempting = true
    // attemptLogin(email, password)
  }

  onSignupButtonPressed = () => {
    // const { name, email, password } = this.state.values
    // const [firstName, lastName] = name.split(' ')
    // const { attemptSignup } = this.props
    // this.isAttempting = true
    // const buildingId = this.selectedBuildingId()
    // const apartmentId = this.selectedApartmentId()
    // attemptSignup({ firstName, lastName, email, password, buildingId, apartmentId })
  }


  onNameChange = (name: string) => {
    this.setState({ name })

    // this.handleChangeFirstName(firstName)
    // this.handleChangeLastName(lastName)
  }

  // handleChangeFirstName = (firstName: string) => {
  //   this.setState({ firstName })
  // }

  // handleChangeLastName = (lastName: string) => {
  //   this.setState({ lastName })
  // }

  handleChangeEmail = (email: string) => {
    const re = /^\S+@\S+$/
    let emailError = null
    if (!re.test(email)) {
      emailError = new Error('You entered an invalid email')
    }
    this.setState({ email, emailError })
  }

  handleChangePassword = (password: string) => {
    let passwordError = null
    if (password.length < 8) {
      passwordError = new Error('Your password is too short')
    }
    this.setState({ password, passwordError })
  }

  // handlePasswordInput

  handleChangePasswordConfirmation = (passwordConfirmation: string) => {
    let confirmationError = null
    if (passwordConfirmation === this.state.password) {
      confirmationError = new Error("Your passwords don't match")
    }
    this.setState({ passwordConfirmation, confirmationError })
  }

  onFbLoginPressed = () => {
    // this.isAttempting = true
    // // this.props.onFbLoginPressed()
    // NavigationActions.buildingSelect({ type: 'replace' })
  }

  disableSigninButton = () => {
    const { email, emailError, passwordError } = this.state
    return (passwordError !== null) && (emailError !== null) && email.length > 0
  }

  onForgotPassword = () => {
    const { onForgotPassword } = this.props
    AlertIOS.prompt(
      'Enter your email',
      null,
      onForgotPassword
    )
  }

  renderLoginContainer = () => {
    const { name, email, password, passwordConfirmation, passwordError, emailError } = this.state
    const { fetching, showingForm, dispatch, loginError, onForgotPassword, navigation, signIn } = this.props

    return (
      <View behavior='height'>
          <LoginForm
            signIn={signIn}
            navigation={navigation}
            fetching={fetching}
            email={email}
            emailError={emailError}
            passwordError={passwordError}
            submitError={loginError}
            password={password}
            disabled={this.disableSigninButton()}
            handleChangeEmail={this.handleChangeEmail}
            handleChangePassword={this.handleChangePassword}
            onLoginButtonPressed={this.onLoginButtonPressed}
            onForgotPassword={this.onForgotPassword}
            onCancelButtonPressed={() => dispatch(LoginActions.showButtonsForm())}
            onSignupButtonPressed={() => this.props.navigation.navigate("SignUp")} />
      </View>
    )
  }

  render () {
    const { isLoggedIn } = this.props
    return (
      <TouchableHighlight
        onPress={() => Keyboard.dismiss()}
        activeOpacity={1}
      >
        <Image
          resizeMode='cover'
          source={Images.loginBackground}
          style={styles.backgroundImage} >
          <KeyboardAwareScrollView
            style={{ backgroundColor: 'transparent' }}
            resetScrollToCoords={{ x: 0, y: 0 }}
            contentContainerStyle={styles.loginContainer}
            extraScrollHeight={80}
            scrollEnabled
          >
            <View style={[styles.topLogo, this.state.topLogo]}>
              <Image source={Images.logo} style={styles.tntLogo} />
              <Image source={Images.name} style={styles.tntName} />
            </View>
            <View style={styles.content}>
              {
                this.renderLoginContainer()
              }
            </View>
          </KeyboardAwareScrollView>
        </Image>
      </TouchableHighlight>
    )
  }
}

export const mapStateToProps = (state) => ({
  fetching: false,
  showingForm: 'LOGIN',
  loginError: false,
  isLoggedIn: false,
})

const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(LoginScreen)
