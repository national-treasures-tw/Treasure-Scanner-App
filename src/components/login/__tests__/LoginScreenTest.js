import React from 'react'
// import { Provider } from 'react-redux'
import { LoginScreen, mapStateToProps } from '../LoginScreen'
// import thunk from 'redux-thunk'
import renderer from 'react-test-renderer'
// import { INITIAL_STATE as login } from '../LoginRedux'
// import configureMockStore from 'redux-mock-store'
// import eos from '../../Services/FixtureApi'
// import facebook from '../../Services/FacebookFixtureApi'

// const middlewares = [ thunk.withExtraArgument({ eos, facebook }) ]
// const mockStore = configureMockStore(middlewares)

const buildings = [
  {
    'id': 1,
    'name': 'Medieval Castle',
    'gallery': null,
    'description': 'DEssfsf',
    'websiteUrl': 'www.killwitch.com',
    'buildingImage': 'http://loremflickr.com/480/640/building',
    'createdAt': '2017-05-12T17:57:37.752Z',
    'updatedAt': '2017-05-24T21:11:22.872Z',
    'LocationId': 1,
    'Apartments': [
      {
        'id': 1,
        'floor': 93,
        'number': '9301',
        'createdAt': '2017-05-12T17:57:37.821Z',
        'updatedAt': '2017-05-12T17:57:37.821Z',
        'BuildingId': 1
      }
    ],
    'Location': {
      'id': 1,
      'streetNumber': '73',
      'streetName': 'Boatworks Drive',
      'apartment': null,
      'city': 'Bayonne',
      'state': 'NJ',
      'country': null,
      'zipcode': 7002,
      'formattedAddress': '73 Boatworks Dr, Bayonne, NJ 07002, USA',
      'lat': 40.6574459,
      'lng': -74.1350035,
      'isPrimary': null,
      'createdAt': '2017-05-10T16:49:04.347Z',
      'updatedAt': '2017-05-26T18:37:54.343Z'
    }
  }
]

const getBuildings = () => buildings

test('dummy component exists', () => {
  const dummy = renderer.create(<LoginScreen getBuildings={getBuildings} />).toJSON()
  expect(dummy).toMatchSnapshot()
})

test('map state to props sets fetching', () => {
  const login = {
    fetching: true
  }
  const startup = {
    initialized: false
  }
  const buildingList = {
    buildings
  }
  const props = mapStateToProps({ login, startup, buildingList })
  expect(login.fetching).toBe(props.fetching)
  expect(login.fetching).not.toBe(props.initialized)
})

/* test('map state to props sets showing form', () => {
  // TODO: this will be replaced with navigation or modals
  const login = {
    showingForm: 'LOGIN'
  }
  const props = mapStateToProps({ login })
  expect(login.showingForm).toBe(props.showingForm)
})

test('map state to props sets is logged in', () => {
  const login = {
    username: 'daredevil'
  }
  const props = mapStateToProps({ login })
  expect(props.isLoggedIn).toBe(true)
})

test('maps dispatch to props', () => {
  const store = (state) => ({
    dispatch: () => {}
  })

  const wrapper = renderer.create(
    <Provider store={store}>
      <Container />
    </Provider>
  )

  const container = wrapper.find(Container)

  expect(container.props.dispatch).toBe(store.dispatch)
})

test('maps dispatch to props sets attemptLogin', () => {
  const store = (state) => ({
    attemptLogin: () => {}
  })

  const wrapper = renderer.create(
    <Provider store={store}>
      <Container />
    </Provider>
  )

  const container = wrapper.find(Container)

  expect(container.props.attemptLogin).toBe(store.attemptLogin)
})

test('container is connected ', () => {
  const store = mockStore({ login })

  const wrapper = renderer.create(
    <Provider store={store}>
      <Container />
    </Provider>
  )

  const container = wrapper.find(Container)

  expect(container.name()).toBe('Connect(LoginScreen)')
}) */
