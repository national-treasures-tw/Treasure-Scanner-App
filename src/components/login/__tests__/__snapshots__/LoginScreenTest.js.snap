// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dummy component exists 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
      },
      undefined,
    ]
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <Image
    resizeMode="cover"
    source={1}
    style={
      Object {
        "backgroundColor": "#292929",
        "height": 1334,
        "left": 0,
        "right": 0,
        "top": 0,
        "width": 750,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "marginTop": 64,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignSelf": "center",
              "backgroundColor": "rgba(0,0,0,0)",
              "flex": 0.18,
              "justifyContent": "center",
              "padding": 20,
            },
            Object {
              "width": 750,
            },
          ]
        }
      >
        <Image
          source={1}
          style={
            Object {
              "alignSelf": "center",
              "height": 54,
              "justifyContent": "center",
              "width": 75,
            }
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "alignItems": "center",
              "color": "#e78720",
              "fontFamily": "Futura",
              "fontSize": 28,
              "justifyContent": "center",
              "textAlign": "center",
            }
          }
        >
          Lifestyle App
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "alignSelf": "center",
            "flex": 0.82,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginTop": 20,
            "paddingTop": 20,
          }
        }
      >
        <View
          behavior="height"
        />
      </View>
    </View>
  </Image>
</View>
`;
